### DISCUSSION TITLE
self hosted docker documentation

### DISCUSSION THREAD

**Discussion by Ganji00** [outline/outline#6810] [Ideas]
Reading through documentation on docker installation there is a mention of [.env.sample](https://github.com/outline/outline/blob/main/.env.sample) file, right now, you need to change manually the URL's of databases in order to make outline work withing a container. I would suggest to add an other samle .env file that would be plug and play ready for docker compose.

**Comment by tommoor** [comment]
What changes would you make to the existing sample?

**Reply by Ganji00** [reply]
So both, postgres and redis URL's should be refering internal docker network names instead of localhost:
`DATABASE_URL=postgres://user:pass@localhost:5432/outline` -> `DATABASE_URL=postgres://user:pass@postgres:5432/outline` and
`REDIS_URL=redis://localhost:6379` -> `REDIS_URL=redis://redis:6379`

**Reply by tommoor** [reply]
It depends on your networking setup I suppose. On Docker for Mac the example values work just fine out of the box

**Reply by Ganji00** [reply]
@tommoor, i dont think so, you can read about it [here](https://docs.docker.com/compose/networking/)

**Reply by Ganji00** [reply]
this is an error, keeping localhost will result to:
```
outline-1       | (node:28) NOTE: We are formalizing our plans to enter AWS SDK for JavaScript (v2) into maintenance mode in 2023.
outline-1       |
outline-1       | Please migrate your code to use AWS SDK for JavaScript (v3).
outline-1       | For more information, check the migration guide at https://a.co/7PzMCcy
outline-1       | (Use `node --trace-warnings ...` to show where the warning was created)
outline-1       | Failed to connect to database
outline-1       | ConnectionRefusedError [SequelizeConnectionRefusedError]: connect ECONNREFUSED 127.0.0.1:5432
outline-1       |     at Client._connectionCallback (/opt/outline/node_modules/sequelize/lib/dialects/postgres/connection-manager.js:131:24)
outline-1       |     at Client._handleErrorWhileConnecting (/opt/outline/node_modules/pg/lib/client.js:327:19)
outline-1       |     at Client._handleErrorEvent (/opt/outline/node_modules/pg/lib/client.js:337:19)
outline-1       |     at Connection.emit (node:events:518:28)
outline-1       |     at Connection.emit (node:domain:488:12)
outline-1       |     at Socket.reportStreamError (/opt/outline/node_modules/pg/lib/connection.js:58:12)
outline-1       |     at Socket.emit (node:events:518:28)
outline-1       |     at Socket.emit (node:domain:488:12)
outline-1       |     at emitErrorNT (node:internal/streams/destroy:169:8)
outline-1       |     at emitErrorCloseNT (node:internal/streams/destroy:128:3)
outline-1       |     at process.processTicksAndRejections (node:internal/process/task_queues:82:21) {
outline-1       |   parent: Error: connect ECONNREFUSED 127.0.0.1:5432
outline-1       |       at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1595:16) {
outline-1       |     errno: -111,
outline-1       |     code: 'ECONNREFUSED',
outline-1       |     syscall: 'connect',
outline-1       |     address: '127.0.0.1',
outline-1       |     port: 5432
outline-1       |   },
outline-1       |   original: Error: connect ECONNREFUSED 127.0.0.1:5432
outline-1       |       at TCPConnectWrap.afterConnect [as oncomplete] (node:net:1595:16) {
outline-1       |     errno: -111,
outline-1       |     code: 'ECONNREFUSED',
outline-1       |     syscall: 'connect',
outline-1       |     address: '127.0.0.1',
outline-1       |     port: 5432
outline-1       |   }
outline-1       | }
outline-1       | Environment configuration is invalid, please check the following:
outline-1       |
outline-1       |
outline-1       | - URL must be a URL address, URL should not be empty
outline-1       | - IFRAMELY_URL must be a URL address
outline-1       | error Command failed with exit code 1.
outline-1       | info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
outline-1 exited with code 1
```

**Comment by matthijskooijman** [comment]
I also ran into this. Even with the mentioned changes, I still get:

    outline_1       | - URL must be a URL address

-> Solved by setting `URL=http://localhost:3000` (for now I am running on localhost for testing)


    outline_1       | - IFRAMELY_URL must be a URL address

This is weird - [documentation says this value is optional](https://docs.getoutline.com/s/hosting/doc/iframely-HwLF1EZ9mo), but apparently it is always validated as an URL even if iframely is disabled by making this empty. For now "solved" by setting `IFRAMELY_URL=http://foo.tld`.

*Edit: Uncommented `IFRAMELY_URL` (as suggested in reply below)is a better way to fix this. Makes some sense, not setting it is different from setting it empty, so maybe the default example env file should have it commented as well*

    outline_1       | {"error":"The server does not support SSL connections","level":"error","message":"The database does not support SSL connections. Set the `PGSSLMODE` environment variable to `disable` or enable SSL on your database server." ,"stack":"SequelizeConnectionError: The server does not support SSL connections\n    at Client._connectionCallback (/opt/outline/node_modules/sequelize/lib/dialects/postgres/connection-manager.js:149:20)\n    at Client._handleErrorWhileConnecting (/opt/outline/node_modules/pg/lib/client.js:327:19)\n    at Client._handleErrorEvent (/opt/outline/node_modules/pg/lib/client.js:337:19)\n    at Connection.emit (node:events:518:28)\n    at Connection.emit (node:domain:488:12)\n    at Socket.<anonymous> (/opt/outline/node_modules/pg/lib/connection.js:77:23)\n    at Object.onceWrapper (node:events:633:26)\n    at Socket.emit (node:events:518:28)\n    at Socket.emit (node:domain:488:12)\n    at addChunk (node:internal/streams/readable:559:12)\n    at readableAddChunkPushByteMode (node:internal/streams/readable:510:3)\n    at Readable.push (node:internal/streams/readable:390:5)\n    at TCP.onStreamRead (node:internal/stream_base_commons:190:23)"}

-> Solved by setting `PGSSLMODE=disable`

**Reply by Ganji00** [reply]
had the same, commenting the `IFRAMELY_URL` out helped.

**Reply by matthijskooijman** [reply]
> Edit: Uncommented IFRAMELY_URL (as suggested in reply below)is a better way to fix this. Makes some sense, not setting it is different from setting it empty, so maybe the default example env file should have it commented as well

I've submitted https://github.com/outline/outline/pull/7401 to fix this in the default config.

**Comment by matthijskooijman** [comment]
~~Also, the docs document to run `docker compose ...` to create/migrate the database, but 1) that should be `docker-compose` and 2) seems no longer needed (the postgres docker container already creates an empty database and the outline docker container automatically migrates on startup AFAIU).~~ *Edit: Turns out this is just new docker syntax, I was still running a bit older docker.*

**Reply by tommoor** [reply]
`docker compose` has been the preferred syntax for several versions. `docker-compose` Is deprecated

**Reply by Ganji00** [reply]
@tommoor is right, docker-compose is deprecated.

**Reply by matthijskooijman** [reply]
Ah, but `docker compose` has not been supported for that long it seems, my Ubuntu 23.10 docker version does not know it yet. But fair to use the new syntax in the docs indeed, sorry for the noise :-)

**Reply by Ganji00** [reply]
Try arch linux 😊
